File: web/src/snapshot.ts
/**
 * The wire shape the telemetry socket pushes — the mirror of engine/snapshot.py.
 *
 * The console is fed whole-state snapshots, never deltas: every message is the
 * complete truth, so there is no reducer here to drift out of step with the
 * engine. `schema` is how a console notices it is older than the engine it is
 * talking to instead of quietly rendering nonsense.
 */

/** 3: the call view gained its live progress — attempt, node, puzzle (#37). */
export const SNAPSHOT_SCHEMA_VERSION = 3;

export type Mode = "tweeted" | "puzzle" | "roguelike";
/**
 * How a Call Session ended, as the store records it.
 *
 * `dropped` is the one no mode handler returns: the engine ended the call
 * itself after an exception, with the caller still on the line. It is kept out
 * of `hangup` so a failure of ours is never counted as a caller walking away
 * (#50).
 */
export type Outcome = "succeed" | "fail" | "exile" | "hangup" | "dropped";

/**
 * Where a Call Session has got to.
 *
 * The terminal states are kept apart deliberately. `handed_off` is a **win**:
 * the channel has left the Call Engine for the success dialplan, so everything
 * after it — the Upstairs Phone ringing, the Operator answering, the pizza —
 * is invisible from here. If a win rendered like a hangup the Operator would
 * learn to distrust the panel, so the two never share a look.
 */
export type CallState =
  | "answering"
  | "in_mode"
  | "handed_off"
  | "exiled"
  | "hung_up"
  | "dropped";

export const TERMINAL_STATES: ReadonlySet<CallState> = new Set<CallState>([
  "handed_off",
  "exiled",
  "hung_up",
  "dropped",
]);

/** What the booth is set to right now (Global Config). */
export interface ConfigView {
  mode: Mode;
  code: string;
  /**
   * The booth's Attempt Limit *now*. Not the one the live caller is judged
   * against — that is `CallView.attempt_limit`, off their frozen Config
   * Snapshot, and the two differ for the length of any call in progress when
   * an Operator changes the setting.
   */
  attempt_limit: number;
  upstream_extension: string;
}

/** Where the caller has got to in the Roguelike Phone-Tree. */
export interface NodeView {
  /** Index into a tree regenerated for this Call Session — meaningless elsewhere. */
  index: number;
  /** Rooms walked through. The readable half of a position. */
  depth: number;
  /** The leaf: where the Code is read aloud. */
  terminal: boolean;
}

/** The live Call Session, as much of it as the engine knows so far. */
export interface CallView {
  session_id: string;
  state: CallState;
  mode: Mode | null;
  caller_id: string | null;
  /** When the call was picked up. The browser advances the clock from here. */
  started_at: string;
  /** `null` while the call is live; the moment it ended once it is over. */
  ended_at: string | null;
  /** The most recent digits dialled, oldest dropped past the engine's cap. */
  digits: string;
  /**
   * Which attempt the caller is on right now, or `null` before the first.
   *
   * Distinct from `attempts`, which is the *final* count and stays 0 until the
   * mode handler returns. One is shown during the call and the other after it.
   */
  attempt: number | null;
  /** The Attempt Limit **this call** is judged against, off its Config Snapshot. */
  attempt_limit: number | null;
  /** `null` unless this is a Roguelike Phone-Tree session. */
  node: NodeView | null;
  /** The riddle drawn from the Puzzle Pool; `null` outside Audio Puzzle Mode. */
  puzzle_id: string | null;
  attempts: number;
  outcome: Outcome | null;
}

export interface Snapshot {
  schema: number;
  config: ConfigView;
  /** `null` is the idle marker: the booth is up, nobody is on the phone. */
  call: CallView | null;
}

export const MODE_LABELS: Record<Mode, string> = {
  tweeted: "Tweeted",
  puzzle: "Audio Puzzle",
  roguelike: "Roguelike Phone-Tree",
};

/** How each state reads on the panel: its headline, and what it actually means. */
export const STATE_COPY: Record<CallState, { label: string; note: string }> = {
  answering: {
    label: "Answering",
    note: "The channel is up. Taking the Config Snapshot this call is judged against.",
  },
  in_mode: {
    label: "On the line",
    note: "The caller is in the game.",
  },
  handed_off: {
    label: "Handed Off",
    note:
      "They won. The channel has left the Call Engine for the success dialplan — " +
      "the Upstairs Phone is ringing, and everything after that is off this panel.",
  },
  exiled: {
    label: "Exiled",
    note: "They burned the Attempt Limit and heard the Exile message.",
  },
  hung_up: {
    label: "Hung up",
    note: "The call ended without a win. Nobody is on the line.",
  },
  dropped: {
    label: "Engine dropped the call",
    note: "The engine ended this call itself — not the caller. Check the engine log.",
  },
};

File: engine/snapshot.py
"""The telemetry snapshot the Operator Console is rendered from (#35, #36).

The console is fed *whole-state* snapshots, never deltas: the Global Config the
booth is set to, plus the live Call Session if one is in progress — or an idle
marker when none is. One builder turns the two in-memory objects into a
JSON-ready dict, so the WebSocket server (and any future consumer) has exactly
one place that knows the wire shape. The ``schema`` number lets a future
console tell an old shape from a new one without guessing.

The live call carries ``started_at`` and (once over) ``ended_at``, never a
duration: the browser advances the elapsed clock itself, so a call on the line
does not generate a message per second purely to tick a timer (ADR-0003).

Note the two Attempt Limits, which are not the same number. The one under
``config`` is Global Config — what the booth is set to *now*. The one under
``call`` came off this call's frozen Config Snapshot, and is what the caller on
the line is actually being judged against. They differ for the length of any
call in progress when an Operator changes the setting, and the Console must
never show the first in place of the second.
"""

from __future__ import annotations

from typing import Any

from core.config import ConfigSnapshot
from core.types import Mode
from engine.call_session import CallSession

# 2: the live Call Session gained its state vocabulary, the digits the caller is
#    dialling, and the moment a finished call ended (#36).
# 3: …and its live progress — which attempt of how many, which room of the maze,
#    which riddle — off the CallObserver seam (#37).
# 4: a call that ended without the mode handler returning now carries an outcome
#    too — ``hangup`` for a caller who hung up mid-call, and the new ``dropped``
#    for an engine failure (#50). ``call.outcome`` was always null for those
#    before, and ``dropped`` is a value no older console has in its union.
SNAPSHOT_SCHEMA_VERSION = 4


def build_snapshot(config: ConfigSnapshot, session: CallSession | None) -> dict[str, Any]:
    """The full console state: Global Config plus the live call, or idle."""
    return {
        "schema": SNAPSHOT_SCHEMA_VERSION,
        "config": _config_view(config),
        "call": _call_view(session),
    }


def _config_view(config: ConfigSnapshot) -> dict[str, Any]:
    return {
        "mode": config.mode,
        "code": config.code,
        "attempt_limit": config.attempt_limit,
        "upstream_extension": config.upstream_extension,
    }


def _call_view(session: CallSession | None) -> dict[str, Any] | None:
    if session is None:
        return None
    mode: Mode | None = session.mode
    if mode is None and session.config is not None:
        # The mode is stamped at pickup; before that the Config Snapshot taken
        # at that moment is the only record of the game this call was given.
        mode = session.config.mode
    return {
        "session_id": session.session_id,
        "state": session.state,
        "mode": mode,
        "caller_id": session.caller_id,
        "started_at": session.started_at.isoformat(),
        # None while the call is live. Present so the Console can stop the
        # elapsed clock where the call stopped rather than at "now".
        "ended_at": session.ended_at.isoformat() if session.ended_at else None,
        "digits": "".join(session.digits),
        # Where the caller is *now*: which attempt of how many they are on.
        # ``attempts`` below is the final count, and stays 0 until the handler
        # returns — the Console shows one during the call and the other after.
        "attempt": session.current_attempt,
        "attempt_limit": _attempt_limit(session),
        "node": _node_view(session),
        "puzzle_id": session.puzzle_id,
        "attempts": session.attempts,
        "outcome": session.outcome,
    }


def _attempt_limit(session: CallSession) -> int | None:
    """The Attempt Limit *this call* is judged against.

    Reported by the flow as each attempt starts, so it is the number actually
    used. Before the first attempt there is nothing to report but the Config
    Snapshot, which is the same number the flow will use — an Operator watching
    a call answer should see "of 3" straight away, not a dash that fills in only
    once the caller dials.
    """
    if session.attempt_limit is not None:
        return session.attempt_limit
    return session.config.attempt_limit if session.config is not None else None


def _node_view(session: CallSession) -> dict[str, Any] | None:
    """Where the caller has got to in the Roguelike Phone-Tree, if anywhere.

    The index is carried for completeness, but ``depth`` is the readable part:
    the tree is regenerated per Call Session, so the index is a coordinate on a
    map only this call has. ``terminal`` is the leaf — where the Code is read
    aloud — which is the one position worth spotting from across a room.
    """
    node = session.node
    if node is None:
        return None
    return {"index": node.index, "depth": node.depth, "terminal": node.terminal}
